src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ProcessedEnvelopeNotificationHandler.java

```diff
@@ -0,0 +1,158 @@
+package uk.gov.hmcts.reform.bulkscanprocessor.tasks;
+
+import com.fasterxml.jackson.core.JsonParseException;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.microsoft.azure.servicebus.ExceptionPhase;
+import com.microsoft.azure.servicebus.IMessage;
+import com.microsoft.azure.servicebus.IMessageHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import uk.gov.hmcts.reform.bulkscanprocessor.exceptions.EnvelopeNotFoundException;
+import uk.gov.hmcts.reform.bulkscanprocessor.exceptions.InvalidMessageException;
+import uk.gov.hmcts.reform.bulkscanprocessor.model.in.msg.ProcessedEnvelope;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.EnvelopeFinaliserService;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.servicebus.MessageAutoCompletor;
+
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+
+public class ProcessedEnvelopeNotificationHandler implements IMessageHandler {
+
+    private static final Logger log = LoggerFactory.getLogger(ProcessedEnvelopeNotificationHandler.class);
+
+    private final EnvelopeFinaliserService envelopeFinaliserService;
+    private final ObjectMapper objectMapper;
+    private final MessageAutoCompletor messageCompletor;
+
+    public ProcessedEnvelopeNotificationHandler(
+        EnvelopeFinaliserService envelopeFinaliserService,
+        ObjectMapper objectMapper,
+        MessageAutoCompletor messageCompletor
+    ) {
+        this.envelopeFinaliserService = envelopeFinaliserService;
+        this.objectMapper = objectMapper;
+        this.messageCompletor = messageCompletor;
+    }
+
+    @Override
+    public CompletableFuture<Void> onMessageAsync(IMessage message) {
+        return CompletableFuture
+            .supplyAsync(() -> tryProcessMessage(message))
```

Comment:
hmm. so this handler tries to exec everything in one go. this is not optimising the feature completable future provides. the way i see it, for instance, this can be extracted step by step making the implementation both easier to read and broken down to relevant stages

```java
// taken from notification one
CompletableFuture
    .supplyAsync(message::getBody, SIMPLE_EXEC)
    .thenApplyAsync(body -> readProcessedEnvelope(body, message.getMessageId(), message.getDeliveryCount()), SERVICE_EXEC)
    .thenApplyAsync(processedEnvelope -> processEnvelope(processedEnvelope, message.getMessageId(), SERVICE_EXEC)
    .thenAcceptAsync(result -> finaliseResult(result, message), SIMPLE_EXEC)
    .handleAsync((voided, throwable) -> { w/e (return future) }, SIMPLE_EXEC)
    .thenComposeAsync(Functions.identity(), ERROR_EXEC)
    .thenRunAsync(() -> log.debug("Error notification consumed. ID {}", message.getMessageId()), SIMPLE_EXEC);
```

each line would represent one piece of action. can pass whole message if need to

Reply:
There are actually two steps:
- process the envelope, i.e. do all the db updates
- finalise the message in the queue (complete/dead-letter)

These are the only ones that take any meaningful amount of time and there's no benefit of splitting them further.

Reply:
one can argue but nevermind. as of wider scope you are right 👍